build: Don't have both a pip and pip-tools requirements files.

feanil opened the pull request:
Previously we would upgrade pip before we upgrade pip-tools.  This
breaks when the latest version of pip is not compatible with the current
version of pip-tools as happened with https://github.com/jazzband/pip-tools/issues/2252

If we re-order the steps so that we upgrade pip-tools first, we know
that this upgrade call will work since it will run with the versions of
pip and pip-compile that ran the last full upgrade.

However in this case the pip.txt file is redundant as the pip-tools.txt
file already has the latest version of pip that is compatible with the
current version of pip-tools being installed.

This changeset also updates the compile-requirements command to ignore
the common_constraints entry for pip so that we can verify the upstream
fix of pip-compile before we remove the entry from
common_constraints.txt upstream.

feanil committed:
build: Don't have both a pip and pip-tools requirements files.

Previously we would upgrade pip before we upgrade pip-tools.  This
breaks when the latest version of pip is not compatible with the current
version of pip-tools as happened with https://github.com/jazzband/pip-tools/issues/2252

If we re-order the steps so that we upgrade pip-tools first, we know
that this upgrade call will work since it will run with the versions of
pip and pip-compile that ran the last full upgrade.

However in this case the pip.txt file is redundant as the pip-tools.txt
file already has the latest version of pip that is compatible with the
current version of pip-tools being installed.

This changeset also updates the compile-requirements command to ignore
the common_constraints entry for pip so that we can verify the upstream
fix of pip-compile before we remove the entry from
common_constraints.txt upstream.

feanil commented:
@timmc-edx  I feel like you've thought about this a bit before so I'd love your thoughts on this change if you have a sec for it.  Though, review from others is also welcome.

timmc-edx commented:
I'm having some trouble seeing the logic here. Won't this still break during a period of incompatible latest/latest? Or is the idea that pip-tools will constrain the pip version somehow and ensure we get a compatible pair?

feanil commented:
The `pip-tools.txt` file has a pip version in it that is compatible with that version of pip-tools. So the initial install of pip-tools should always install co-compatible versions of both.  Then we attempt to upgrade both.

timmc-edx commented:
In a situation where latest of pip and pip-tools are incompatible, I would predict the following:

- We upgrade and install pip-tools.txt, bringing in the max compatible pip version.
- We then upgrade and install pip.txt, bringing in a newer (incompatible) pip.
- We then proceed with pip-compile commands, which break.

What am I missing here?

timmc-edx commented:
...actually, why have pip.txt at all if pip-tools.txt includes it?

feanil commented:
Yea, I was thinking about that too, like do we need both?  Thinking about it more, the pip-compile of `pip-tools.txt` should upgrade both so yea maybe we can just drop that.  Let me make that change.

timmc-edx commented on Makefile:
```diff
@@ -122,10 +121,8 @@ compile-requirements: pre-requirements $(COMMON_CONSTRAINTS_TXT) ## Re-compile *
-	pip-compile -v --allow-unsafe ${COMPILE_OPTS} -o requirements/pip.txt requirements/pip.in
-	pip install -r requirements/pip.txt
 
 	pip-compile -v ${COMPILE_OPTS} -o requirements/pip-tools.txt requirements/pip-tools.in
```
I believe we'll need `--allow-unsafe` in order to get a `pip` dep in the `pip-tools.txt` lockfile (and then run `make compile-requirements` to rebuild). Currently it's excluded.

```suggestion
	pip-compile -v --allow-unsafe ${COMPILE_OPTS} -o requirements/pip-tools.txt requirements/pip-tools.in
```

(Can't test locally, having some issues with GH at the moment.)

feanil replied:
Yea, you're right, good catch.

feanil committed:
build: Update github workflows that relied on pip.txt

Use the `pre-requirements` target to ensure that pip is installed at a
version that's valid for the rest of the workflow.

feanil committed:
chore: Run `make compile-requirements`

timmc-edx reviewed (approved):
Looks good! Just needs updated PR title/desc.
